contracts/protocol/libraries/logic/MarketplaceLogic.sol

```diff
@@ -434,38 +415,106 @@ library MarketplaceLogic {
             return;
         }
 
-        DataTypes.ReserveData storage reserve = reservesData[vars.creditToken];
-        vars.xTokenAddress = reserve.xTokenAddress;
-
-        require(vars.xTokenAddress != address(0), Errors.ASSET_NOT_LISTED);
+        DataTypes.ReserveData storage reserve = ps._reserves[vars.creditToken];
         ValidationLogic.validateFlashloanSimple(reserve);
-        // TODO: support PToken
-        IPToken(vars.xTokenAddress).transferUnderlyingTo(to, vars.creditAmount);
+        IPToken(vars.creditXTokenAddress).transferUnderlyingTo(
+            to,
+            vars.creditAmount
+        );
 
         if (vars.isETH) {
             // No re-entrancy because it sent to our contract address
             IWETH(params.weth).withdraw(vars.creditAmount);
         }
     }
 
+    /**
+     * @notice Flash mint 90% of listingPrice as pToken so that seller's NFT can be traded in advance.
+     * Repayment needs to be done after the marketplace exchange by transferring funds to xTokenAddress
+     * @dev
+     * @param ps The pool storage pointer
+     * @param params The additional parameters needed to execute the buyWithCredit/acceptBidWithCredit function
+     * @param vars The marketplace local vars for caching storage values for future reads
+     * @param seller The NFT seller
+     */
+    function _flashSupplyFor(
+        DataTypes.PoolStorage storage ps,
+        DataTypes.ExecuteMarketplaceParams memory params,
+        MarketplaceLocalVars memory vars,
+        address seller
+    ) internal {
+        if (vars.isETH) {
+            return; // impossible to supply ETH on behalf of the
+        }
+
+        DataTypes.ReserveData storage reserve = ps._reserves[vars.creditToken];
+        DataTypes.UserConfigurationMap storage sellerConfig = ps._usersConfig[
+            seller
+        ];
+        DataTypes.ReserveCache memory reserveCache = reserve.cache();
+        uint16 reserveId = reserve.id; // cache to reduce one storage read
+
+        reserve.updateState(reserveCache);
```

Comment:
Here we `updateState` and `updateInterestRates`, and will do it again in `executeBorrow` while `_handleFlashLoanRepayment`

Reply:
Ok, I'll think about this a bit later